modules/spring-data-ext/src/main/java/org/apache/ignite/springdata/repository/support/IgniteResourceProviderImpl.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.springdata.repository.support;
+
+import org.apache.ignite.Ignite;
+import org.apache.ignite.Ignition;
+import org.apache.ignite.client.IgniteClient;
+import org.apache.ignite.configuration.ClientConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.springdata.proxy.ClosableIgniteClientProxy;
+import org.apache.ignite.springdata.proxy.ClosableIgniteProxy;
+import org.apache.ignite.springdata.proxy.IgniteClientProxy;
+import org.apache.ignite.springdata.proxy.IgniteProxy;
+import org.apache.ignite.springdata.proxy.IgniteProxyImpl;
+import org.apache.ignite.springdata.repository.config.RepositoryConfig;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+
+/** Represents default implementation of {@link IgniteResourceProvider} */
+public class IgniteResourceProviderImpl implements IgniteResourceProvider, ApplicationContextAware, DisposableBean {
```

Comment:
Actually, what is a purpose of this class? Spring closes AutoCloseable beans by default, provides Lazy annotation for lazy initialization, routing by bean name looks arguable (and nevertheless earlier it was implemented different way).

Reply:
The main problem with the previous implementation: Ignite / IgniteClient instances started by Spring Data(in case Ignite / IgniteClient configuration is used to configure the repository. See RepositoryConfing#igniteCfg or igniteSpringCfgPath) are not closed at all. Also Ignite instances are started during bean initialization. Therefore, in some cases, bean initialization can fail with error but started Ignite instances will continue to work. By moving Ignite management logic to the separate bean - IgniteResourceProvider - we can control Ignite instances regardless of the repository lifecycle. It also allows users to use custom logic for obtaining Ignite Instance associated with a repository (see EnableIgniteRepositories#igniteResourceProvider).

Reply:
Is there any issues with managing Ignite as a bean in every case? In case with igniteCfg it is possible to create Ignite bean with method, isn't it?

I mean smth like that:
`public Ignite igniteInstance(@Autowired IgniteConfiguration cfg)`
`public IgniteProxy igniteProxy(@Autowired Ignite ignite)`

Reply:
Starting Ignite with `new` inside Repository looks bad. It is not a Spring way

Reply:
I think it's backward compatibility. Now we already have the functionality to configure a Spring Data repository just providing Ignite/ClientConfiguration bean name for each repository. Please, see RepositoryConfing#igniteCfg or igniteSpringCfgPath.